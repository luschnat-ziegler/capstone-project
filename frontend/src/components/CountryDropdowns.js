import Select from 'react-select'
import styled from 'styled-components/macro'

export default function CountryDropdowns ({countries, handleDisplayedCountries, displayedCountries}) {


    const dropdownOptions = countries.map(country => ({
            value: country, 
            label: country.country
        }))
    
    return (
        <Wrapper>
        <Select styles={styles}
            options = {dropdownOptions}
            onChange = {handlerLeft}
        />
        <Select styles={styles}
            options = {dropdownOptions}
            onChange = {handlerRight}
        />
        </Wrapper>
    )

    // ToDo: Refactor into one function
    function handlerLeft (e) {
        handleDisplayedCountries({
            countryLeft: e.value,
            countryRight: displayedCountries.countryRight
        })
    }

    function handlerRight (e) {
        handleDisplayedCountries({
            countryLeft: displayedCountries.countryLeft,
            countryRight: e.value
        })
    }
}

const Wrapper = styled.div`
    display: flex;
    justify-content: space-between;
    width: 100%;
`

const styles = {
    container: base => ({
      ...base,
      flex: 1
    })
  };