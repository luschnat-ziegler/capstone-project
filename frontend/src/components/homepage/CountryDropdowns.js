import Select from 'react-select'
import styled from 'styled-components/macro'
import PropTypes from 'prop-types'

export default function CountryDropdowns({
  countries,
  handleDisplayedCountries,
  displayedCountries,
}) {
  const dropdownOptions = countries.map((country) => ({
    value: country,
    label: country.country,
  }))

  return (
    <Wrapper>
      <Select
        styles={styles}
        options={dropdownOptions}
        onChange={handlerLeft}
        name="handlerLeft"
        placeholder={'Select Country 1'}
        defaultValue={
          displayedCountries.countryLeft.country && {
            value: displayedCountries.countryLeft,
            label: displayedCountries.countryLeft.country,
          }
        }
      />
      <Select
        styles={styles}
        options={dropdownOptions}
        onChange={handlerRight}
        name="handlerRight"
        placeholder={'Select Country 2'}
        defaultValue={
          displayedCountries.countryRight.country && {
            value: displayedCountries.countryRight,
            label: displayedCountries.countryRight.country,
          }
        }
      />
    </Wrapper>
  )

  function handlerLeft(e) {
    handleDisplayedCountries({
      ...displayedCountries,
      countryLeft: e.value,
    })
  }

  function handlerRight(e) {
    handleDisplayedCountries({
      ...displayedCountries,
      countryRight: e.value,
    })
  }
}

const Wrapper = styled.div`
  display: flex;
  padding: 10px 5px;
  justify-content: space-between;
  width: 100%;
`

const styles = {
  container: (base) => ({
    ...base,
    flex: 1,
  }),
}

CountryDropdowns.propTypes = {
  displayedCountries: PropTypes.object,
  countries: PropTypes.array,
  handleDisplayedCountries: PropTypes.func,
}
