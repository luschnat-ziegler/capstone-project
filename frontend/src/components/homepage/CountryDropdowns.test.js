import { fireEvent, getAllByTestId, render } from '@testing-library/react'
import 'jest-styled-components'
import CountryDropdowns from './CountryDropdowns'

jest.mock('react-select', () => ({ options, value, onChange }) => {
  function handleChange(event) {
    const option = options.find((option) => option.value === event.currentTarget.value)
    onChange(option)
  }
  return (
    <select data-testid="select" value={value} onChange={handleChange}>
      {options.map(({ label, value }) => (
        <option key={value} value={value}>
          {label}
        </option>
      ))}
    </select>
  )
})

const displayedCountriesEmpty = {
  countryLeft: {},
  countryRight: {},
}

const displayedCountriesSelected = {
  countryLeft: {
    id: 18,
    country: 'Bhutan',
    region: 'Asia',
    freedom: 59,
    gender: 49,
    lgbtq: 35,
    environment: null,
    corruption: 75,
    inequality: 65,
    total: 57,
  },
  countryRight: {
    id: 4,
    country: 'Angola',
    region: 'SSA',
    freedom: 32,
    gender: 32,
    lgbtq: 29,
    environment: 89,
    corruption: 19,
    inequality: 30,
    total: 38,
  },
}

const countries = [
  {
    corruption: 6,
    country: 'Afghanistan',
    environment: null,
    freedom: 27,
    gender: 32,
    id: 1,
    inequality: null,
    lgbtq: 16,
    region: 'Asia',
    total: 20,
  },
  {
    corruption: 31,
    country: 'Albania',
    environment: 58,
    freedom: 67,
    gender: 75,
    id: 2,
    inequality: 76,
    lgbtq: 27,
    region: 'Europe',
    total: 56,
  },
]

describe('CountryDropdowns', () => {
  it('renders correctly', () => {
    const handleDisplayedCountries = jest.fn()
    const { container } = render(
      <CountryDropdowns
        countries={countries}
        displayedCountries={displayedCountriesEmpty}
        handleDisplayedCountries={handleDisplayedCountries}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('features two dropdown menus', () => {
    const handleDisplayedCountries = jest.fn()
    const { getByText } = render(
      <CountryDropdowns
        countries={countries}
        displayedCountries={displayedCountriesEmpty}
        handleDisplayedCountries={handleDisplayedCountries}
      />
    )
    const dropdownLeft = getByText(/select country 1/i)
    const dropdownRight = getByText(/select country 2/i)

    expect(dropdownLeft).toBeInTheDocument()
    expect(dropdownRight).toBeInTheDocument()
  })

  it('triggers callback on change', () => {
    const handleDisplayedCountries = jest.fn()
    const { getAllByTestId } = render(
      <CountryDropdowns
        countries={countries}
        displayedCountries={displayedCountriesEmpty}
        handleDisplayedCountries={handleDisplayedCountries}
      />
    )

    const dropdownLeft = getAllByTestId('select')[0]
    fireEvent.change(dropdownLeft)
    expect(handleDisplayedCountries).toHaveBeenCalled()
  })
})
