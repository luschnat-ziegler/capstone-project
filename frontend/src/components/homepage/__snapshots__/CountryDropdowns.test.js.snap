// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountryDropdowns renders correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 10px 5px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
}

<div>
  <div
    class="c0"
  >
    <select
      data-testid="select"
    >
      <option
        value="[object Object]"
      >
        Afghanistan
      </option>
      <option
        value="[object Object]"
      >
        Albania
      </option>
    </select>
    <select
      data-testid="select"
    >
      <option
        value="[object Object]"
      >
        Afghanistan
      </option>
      <option
        value="[object Object]"
      >
        Albania
      </option>
    </select>
  </div>
</div>
`;
